### DISCUSSION TITLE
Add multiple jobs to the schedule asynchronously

### DISCUSSION THREAD

**Discussion by hervesimard** [quartznet/quartznet#649] [General]
I am trying to use Quartz.Net and up to now the documentation and examples have helped me get a good understanding of what it does, however I got a problem and wanted to share it with you.

I encapsulated the Scheduler in a class called ScriptInjector because I wanted to create an instance of a Scheduler which only role consisted of inserting jobs in a database. We wanted to be able to sustain up to 200 000 simultaneous insertions (and more in a principle of scalability).

To test out my approach I did the following:

```
           List<JobHolder> jobs = new List<JobHolder>();
           for (int count = 0; count < lowerBounds; count++)
            {
                IJobDetail job = JobBuilder.Create<SimpleJob>()
                    .WithIdentity("job" + count, "group_1")
                    .RequestRecovery()
                    .Build();

                long timeDelay = (long)(r.NextDouble() * 2500);
                job.JobDataMap.Put(SimpleJob.DelayTime, timeDelay);

                ITrigger trigger = TriggerBuilder.Create()
                    .WithIdentity("Trigger " + count, "group_1")
                    .StartAt(DateBuilder.FutureDate(10000 + count * 100, IntervalUnit.Millisecond))
                    .Build();

                jobs.Add(new JobHolder
                {
                    JobDetail = job,
                    Trigger = trigger
                });
            }

            watch.Start();

            var tasks = jobs.Select(item =>
            {
                return scriptInjector.AddToSchedule(item.JobDetail, item.Trigger);
            });
            
            await Task.WhenAll(tasks);
```

Where the `JobHolder` class is like this:

```
    public class JobHolder
    {
        public IJobDetail JobDetail;

        public ITrigger Trigger;
    }
```

Here's the AddToSchedule method:

```
        public async Task AddToSchedule(IJobDetail jobDetail, ITrigger trigger)
        {
            Console.WriteLine($"Now executing {jobDetail.ToString() }");
            await shed.ScheduleJob(jobDetail, trigger);
            Console.WriteLine("Done.");
        }
```

What I found out is that the jobs were added sequentially. If I tried to launch multiple tasks to do the injection I'd end up having time out exceptions thrown because the locks would not be accessible. Please note that I tried with the dictionary as well and, following the approach I undertook up above, it took about 12 minutes to complete.

So is there a way to run many tasks / processes to add many jobs simultaneously?

**Comment by lahma** [comment]
For such aggressive insert you are probably hitting job store's LockOnInsert option that is on by default.

        /// <summary>
        /// Whether or not to obtain locks when inserting new jobs/triggers.
        /// </summary>
        /// <remarks>
        /// <para>
        /// Defaults to <see langword="true" />, which is safest - some db's (such as
        /// MS SQLServer) seem to require this to avoid deadlocks under high load,
        /// while others seem to do fine without.  Settings this to false means
        /// isolation guarantees between job scheduling and trigger acquisition are
        /// entirely enforced by the database.  Depending on the database and it's
        /// configuration this may cause unusual scheduling behaviors.
        /// </para>
        /// <para>
        /// Setting this property to <see langword="false" /> will provide a
        /// significant performance increase during the addition of new jobs
        /// and triggers.
        /// </para>
        /// </remarks>
        public bool LockOnInsert { get; set; } = true;

So you could try setup you job store configuration with:

       properties["quartz.jobStore.lockOnInsert "] = "false"

But be aware of the documented consequences.

**Comment by hervesimard** [comment]
Here are the properties I declared to instantiate my scheduler:

```
NameValueCollection properties = new NameValueCollection
        {
            ["quartz.scheduler.instanceName"] = "TestScheduler",
            ["quartz.scheduler.instanceId"] = "instance_one",
            ["quartz.threadPool.type"] = "Quartz.Simpl.SimpleThreadPool, Quartz",
            ["quartz.threadPool.threadCount"] = "10",
            ["quartz.jobStore.misfireThreshold"] = "60000",
            ["quartz.jobStore.type"] = "Quartz.Impl.AdoJobStore.JobStoreTX, Quartz",
            ["quartz.jobStore.useProperties"] = "false",
            ["quartz.jobStore.dataSource"] = "default",
            ["quartz.jobStore.tablePrefix"] = "QRTZ_",
            ["quartz.jobStore.clustered"] = "true",
            ["quartz.jobStore.lockOnInsert"] = "false",
            ["quartz.jobStore.driverDelegateType"] = "Quartz.Impl.AdoJobStore.SqlServerDelegate, Quartz",
            ["quartz.dataSource.default.connectionString"] = "Server=***;Database=Quartz;User Id=***;Password=***",
            ["quartz.dataSource.default.provider"] = "SqlServer",
            ["quartz.serializer.type"] = "binary"
        };
```

Have I done something wrong?

**Comment by windsOne** [comment]
@hervesimard   I have the same needs as you, can you communicate with each other? thanks

**Comment by lahma** [comment]
You can also use the batch-oriented API for add the jobs, here's a version for that which imports 1000 jobs at a time. I'll check if there's anything to easily improve here.

```
            var jobs = new Dictionary<IJobDetail, IReadOnlyCollection<ITrigger>>(1_000);
            for (int count = 0; count < lowerBounds; count++)
            {
                IJobDetail job = JobBuilder.Create<SimpleJob>()
                    .WithIdentity("job" + count, "group_1")
                    .RequestRecovery()
                    .Build();

                long timeDelay = (long) (r.NextDouble() * 2500);
                job.JobDataMap.Put(SimpleJob.DelayTime, timeDelay);

                ITrigger trigger = TriggerBuilder.Create()
                    .WithIdentity("Trigger " + count, "group_1")
                    .StartAt(DateBuilder.FutureDate(10000 + count * 100, IntervalUnit.Millisecond))
                    .Build();

                jobs.Add(job, new List<ITrigger> { trigger});

                if (jobs.Count == 1000)
                {
                    await shed.ScheduleJobs(jobs, false).ConfigureAwait(false);
                    jobs.Clear();
                    Console.WriteLine($"{watch.Elapsed}: processed {(double) count/lowerBounds:P1}");
                }
            }
```

**Comment by lahma** [comment]
When you import with the ScheduleJobs batch API you shouldn't need the LockOnInsert hack either. To get more speed out of this probably some bigger in changes need to be done, probably in 4.0 timeframe. As you can see, my sample doesn't multi-thread, if multi-threading then you would again need the LockOnInsert thing. See what becomes the bottleneck, with current logic SQL Server is hammered quite a bit.

As you are serializing job data, you could also try out the JSON serialization feature, it should be a bit faster.

I've tweaked SQL generation a bit for 3.0.7 release which should save some CPU cycles, but not that much for your use case.

Hope you can get the speed you need with these tips.
